src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+            tail.prev.next = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        size++;
+        isValidIndex(index);
+
+        Node<T> newNode = new Node<>(null, value, null);
+
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size - 1) {
+            getNodeByIndex(size - 2).next = newNode;
+            newNode.prev = getNodeByIndex(size - 2);
+            tail = newNode;
+        } else {
+            Node<T> prevNode = getNodeByIndex(index - 1);
+            newNode.next = prevNode.next;
+            newNode.prev = prevNode;
+            prevNode.next.prev = newNode;
+            prevNode.next = newNode;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isValidIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isValidIndex(index);
+        Node<T> currentNode;
+        T oldVal;
+        if (index == size) {
+            currentNode = tail;
+            currentNode.value = value;
+            oldVal = currentNode.value;
+            return oldVal;
+        }
+        currentNode = getNodeByIndex(index);
+        oldVal = currentNode.value;
+        currentNode.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isValidIndex(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> currentNode = head; currentNode != null; currentNode = currentNode.next) {
+            if ((object == null && currentNode.value == null)
+                    || (object != null && object.equals(currentNode.value))) {
+                unlink(currentNode);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index <= size / 2) {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            return current;
+        } else {
+            Node<T> current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+            return current;
+        }
+    }
+
+    private void isValidIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private T unlink(Node<T> x) {
```

Comment:
we should not name anything with just a single letter, and please be more attentive in fixing the comments me and Andrii left quite a few regarding the naming and all of them are very similar. So if you receive a comment try and look up similar cases and fix them as well to avoid facing the same problem each review